src/Streams.jl

```diff
@@ -282,14 +280,20 @@ function Base.unsafe_read(http::Stream, p::Ptr{UInt8}, n::UInt)
     nothing
 end
 
-function Base.readbytes!(http::Stream, buf::IOBuffer, n=bytesavailable(http))
-    Base.ensureroom(buf, n)
-    unsafe_read(http, pointer(buf.data, buf.size + 1), n)
+@noinline bufcheck(buf, n) = ((buf.size + n) <= length(buf.data)) || throw(ArgumentError("Unable to grow response stream IOBuffer large enough for response body size"))
+
+function Base.readbytes!(http::Stream, buf::Base.GenericIOBuffer, n=bytesavailable(http))
+    Base.ensureroom(buf, buf.size + n)
+    # check if there's enough room in buf to write n bytes
+    bufcheck(buf, n)
+    data = buf.data
+    GC.@preserve data unsafe_read(http, pointer(data, buf.size + 1), n)
```

Comment:
_Sleepy drive by review alert_

Should this be `Base.ensureroom(buf, buf.ptr + n - 1)` and `pointer(data, buf.ptr)` and similarly for `bufcheck`?

Reply:
Well, perhaps, but we're updating `buf.size` ourselves (and not `buf.ptr`), so `buf.size` is correct for this case. From what I can tell, `buf.size` tracks the last byte _written_ to the IOBuffer, whereas `buf.ptr` tracks the next by to _read_ from the IOBuffer. So if that's right, then I think tracking `buf.size` is correct here.

Reply:
AFAIK, writing to IOBuffer should advance both `ptr` and `size`:
```julia
julia> io = IOBuffer(UInt8[1, 2, 3, 4, 5, 6], write=true)
IOBuffer(data=UInt8[...], readable=false, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1)

julia> write(io, "aaaa")
4

julia> io
IOBuffer(data=UInt8[...], readable=false, writable=true, seekable=true, append=false, size=4, maxsize=Inf, ptr=5, mark=-1)

julia> seekstart(io)
IOBuffer(data=UInt8[...], readable=false, writable=true, seekable=true, append=false, size=4, maxsize=Inf, ptr=1, mark=-1)

julia> write(io, "aa")
2

julia> io
IOBuffer(data=UInt8[...], readable=false, writable=true, seekable=true, append=false, size=4, maxsize=Inf, ptr=3, mark=-1)
```
Note that seekstart will allow us reuse the parts of the buffer that were pre-allocated earlier.